### PR TITLE
Implement pull module code action

### LINKED ISSUE #30757: Implement Pull module Code Action
**Description:**
$title

In order to implement this particular use-case, the Language Server has to either depend on executing the Ballerina pull command internally, or to use a pull API provided via the developer tools

### LINKED ISSUE #31370: Couldn't get the ballerinax stdlib modules in the auto Completions
**Description:**
$subject. See the following images for more info.

Screenshot 2021-06-22 at 11 14 46 AM

Screenshot 2021-06-22 at 11 16 23 AM

This issue gets due to the ballerinax modules are not in the distribution. But this issue didn't solve even after running `bal run` command.

Affected Versions:
SLBeta1

### PR DESCRIPTION
**Pull request by IMS94**
## Purpose
$subject

Fixes #30757
Fixes #31370

https://user-images.githubusercontent.com/11285165/128016353-fb2bfa77-27e9-4c13-b9db-238683a30fde.mp4

## Approach
Added a new background task service to run background tasks. Implemented pull module code action to use that to submit pull tasks and notify vscode, etc.

VSCode plugin side was updated to save all files before invoking the pull module command.

Related vscode plugin PR: https://github.com/wso2-enterprise/ballerina-plugin-vscode/pull/60

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/imports/PullModuleCodeAction.java [hunk 1]
```diff
@@ -45,71 +45,50 @@
...
-            action.setKind(CodeActionKind.QuickFix);
-            action.setCommand(new Command(commandTitle, PullModuleExecutor.COMMAND, args));
-            action.setDiagnostics(CodeActionUtil.toDiagnostics(diagnostics));
-            return Collections.singletonList(action);
-        }
-        return Collections.emptyList();
+        String commandTitle = CommandConstants.PULL_MOD_TITLE;
+        CodeAction action = new CodeAction(commandTitle);
+        action.setKind(CodeActionKind.QuickFix);
+        action.setCommand(new Command(commandTitle, "ballerina.packages.pull", args));
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/imports/PullModuleCodeAction.java:85]: We can reuse the PULL_MODULE_COMMAND const here

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/task/BackgroundTaskService.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+     */
+    public boolean isRunning(PackageId packageId, String taskName) {
+        TaskKey key = TaskKey.from(packageId, taskName);
+        return tasks.containsKey(key) && !tasks.get(key).isDone();
+    }
+
+    /**
+     * Cancels all running tasks and clear tasks map.
+     */
+    public void shutdown() {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/task/BackgroundTaskService.java:106]: Will we have the requirement of cancelling an individua task? if not we can add it later when required

**Review by malinthar — APPROVED** [review]
Approved with minor suggestions

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/task/BackgroundTaskService.java (hunk 2 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/task/BackgroundTaskService.java:106]: We will have such a requirement in the future. Even for pull module code action, we should provide the ability to cancel if it takes longer. I have added to `todo` for that. Let's add that API when we implement the cancellation support.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/imports/PullModuleCodeAction.java (hunk 1 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/imports/PullModuleCodeAction.java]: Yes. Missed it. Done!

**Review by malinthar — DISMISSED** [review]
LGTM

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java [hunk 3]
```diff
@@ -98,7 +98,7 @@
 
     public static final String CREATE_INITIALIZER_TITLE = "Create initializer";
 
-    public static final String PULL_MOD_TITLE = "Pull from Ballerina Central";
+    public static final String PULL_MOD_TITLE = "Pull missing module(s)";
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java:101]: What if we use `unresolved modules` ?

Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java:101]: +1

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Pull request merged**